### DISCUSSION TITLE
Thank you Kang Yu it feels like you just saved my life and sanity

### DISCUSSION THREAD

**Discussion by koto9x** [kangyu-california/PersistentWindows#77] [General]
I was looking for a solution to my monitor problem forever. I tested yours and it helps. Thanks i want to buy you a coffee but can't find you on Twitter or anywhere else. Hope you have a good day.

**Comment by kangyu-california** [comment]
If PW works for you, give it a star is good enough

**Comment by rayjay1359** [comment]
I agree it works perfectly for me
thank you

i don't know how to give it a star or I would

**Comment by Naton-1** [comment]
+1 to add a thank you very much!!

**Comment by daktfi** [comment]
+1 here, too! Many thanks, really great tool! As OP said - saved life and sanity.
I have a second monitor on notebook, and when I try to put it to sleep - the monitor wakes up way after main screen and win10 drags all the windows to primary.
The only flaw (and I don't think it's PW's one) is that I have a very custom taskbars layout: SECONDARY taskbar is on the right on MAIN screen, and PRIMARY taskbar is on the bottom of extra monitor that is turned into portrait mode. Yep, it's quite weird, but I have my reason. The thing is, when I disconnect second monitor - primary taskbar goes to main screen down, and when I reconnect second monitor - it goes back to the second monitor, but BOTH taskbars are on the right side.
I don't know if taskbar is a window good enough to be rearranged by PW, but if so, can you please fix it as well?

**Reply by kangyu-california** [reply]
PW is capable of restoring primary taskbar (to the correct screen edge)
There are two requirements mentioned in Readme
1. the taskbar need be unlocked (i.e. free to drag by mouse)
2. do not move mouse during auto recovery

**Reply by daktfi** [reply]
Thanks a lot for clarification! Will check that out.
Anyway, thanks again - this app does wonders!

**Reply by daktfi** [reply]
Tested it just now.
Layout A: single monitor (notebook's primary landscape), single (primary) taskbar on the right.
Layout B: two monitors (NB's primary landscape, secondary portrait), two taskbars, primary on the bottom of the second monitor, secondary taskbar on the right edge on the primary monitor (they both on shorter side to minimize lost space).
1. Disconnecting second monitor works as expected, primary taskbar goes to the right edge of primary (now the only) monitor.
2. Reconnecting second monitor back is a bit flawed: secondary taskbar goes to the right edge of primary monitor (check!), but primary taskbar goes to the right edge of secondary monitor (expected to go to bottom of secondary monitor). That's probably because I place taskbars on different edges of respective screens, dunno if many people do that.
Anyway, that's a WHOLE LOT BETTER than it is w/o your app! Can't thank you enough for it!

UPD: Tried disabling the checkbox "taskbar on all monitors" to keep it to single primary taskbar. 
Layout A: single monitor, taskbar on the right.
Layout B: two monitors, landscape+portrait, taskbar on secondary's bottom.
1. Disconnect: taskbar goes to the single monitor's right edge (check!).
2. Reconnect: taskbar goes to the secondary monitor's right edge (expected bottom).
Hope this additional testing helps. If you need any more tests - just tell me what you want tested, I'll be glad to help any way I can!

UPD2: sorry, I just realized that all this time I was talking about "right" side when in fact it was left. Weird. Hope, this doesn't matter program-wise (it probably shouldn't).

**Reply by kangyu-california** [reply]
Thanks for the feedback. I tried your configuration but could not reproduce the issue.

My setup is : Windows 10 21H1 OS build 19043.1348, dual monitor, one monitor in landscape mode with taskbar on left or right side, the other monitor is in portrait mode with taskbar at bottom.

Windows seems doing a good job without messing up taskbar positions, hence PW only need to restore a few standard windows.

Correct me if it matters, there is no such term as primary/secondary taskbar, user can only specify primary/secondary monitor.

**Reply by daktfi** [reply]
My setup:
Gigabyte Aorus 15G notebook (there can be some hardware issues, how unlikely that sounds). 
Windows 10 Home, 21H2, OS build 19044.1387. Dunno what's this, but for completeness of it: Windows Feature Experience Pack 120.2212.3920.0

Primary/secondary taskbar is a bit complicated. It's never called like this, however, you can have only ONE taskbar that has it's notification area filled with, right, notifications. You can have multiple sets of quick-launch icons and apps icons (I believe these can be switched on/off), but only ONE notification are (like where Skype, WiFi and the like reside). So, this is the taskbar I call "primary". The idea is that even running full-screen apps on primary monitor I'll be able to see notification on secondary monitor. 

Landscape/portrait are a bit of enforced (my workplace is a bit cramped and second monitor just don't quite fit other way), and a bit of a acquired taste - it's quite good for reading longer texts. Placing taskbar to shorter screen edge is a Linux habit of mine (programming, etc). Anyway, that's not important.

So, when I have "Show taskbar on all displays" enabled and have more than one monitor Windows puts "primary" taskbar (as defined above, with notification area) on the primary monitor and "secondary" (w/o notification area) on others. I can switch these manually, but after disconnecting and reconnecting monitors (I have to do that a few times a week working from other places) Windows puts 'em the way it like 'em, not the way I like 'em. 

I just did a bit of testing again, seems I need to go reboot as behavior became non-reproducible, will do that in a few hours, too busy right now.

Anyway, besides taskbar flaw (which I pretty much got used to already) your program does a real great work of placing all other windows correctly, and that's a huge help - I only have to adjust position of taskbar that's already on the right monitor...

Well, thank you again! Did you ever think about propose your app to be included in Microsoft PowerToys package? I think it will be a great addition to them! Well, I think your app MUST be a part of Windows settings, only MS didn't did it, and PowerToys is second best place.

UPDATE. Sorry for delay. Rebooted and repeatedly dis-/re-connected secondary display with single and multiple taskbars.
When "Show taskbar on all displays" is enabled, correct taskbar is correctly placed on primary monitor (to the side, primary if one monitor, secondary if two). On the secondary monitor correct taskbar is moved, but it is attached to the same side it was on single monitor, and doesn't move to the bottom.
When "Show taskbar on all displays" is disabled, taskbar is moved between displays as expected, on DISconnect it is correctly placed to the side, but on REconnect it stays on the side instead of going to bottom.

Looking at the EventViewer I see a whole lot of events 9990 (and a few 9999). Basically they're all the same:
1. Display layout changed between Display_Loc0x0_Res1920x1080__Display_Loc1920xM510_Res1080x1920 and Display_Loc0x0_Res1920x1080.
2. Start restoring window layout back to {timestamp} {layout}.
3. Restore finished in pass 2 with {N} windows recovered for display settings {layout}.

No other info there.
A few 9999 (basically "nothing to do") are after 9990 "Display layout changed to " same layout, I believe it was from the time of system boot (when Windows discovered same display several times).

I have nVidia driver version 510.06 installed (not yet current, remnants of experiments to improve WSL GUI support), not sure if that matters. Will revert to current one (497.09) later today to be sure. Intel graphics driver is current as far as I can tell, will check that too.

UPDATE2. Changed driver to 497.09, same behavior: on reconnecting second display taskbar moves to the side. Can it be because I have it double height when it's on the bottom of secondary monitor? When I use "Restore windows from disk" the taskbar on secondary display is standard height (single row), but I especially captured with two-taskbar.

Additional test:
When I use "Restore windows from disk" after reconnecting monitor - taskbar goes to bottom and is single-row.
When I dragged taskbar manually to the side of secondary monitor and then used "Restore windows from disk" the taskbar was moved to bottom and was two-rows.
Seems something there's worth further investigation. How can I help with that? Events are not very informative, unfortunately. Is there some "extended debug log" switch?

**Reply by kangyu-california** [reply]
Thanks for the new findings and offering help.
You may want to install visual studio (2017 or above) and build a debug version.
The function MoveTaskbar() in PersistentWindowProcessor.cs is the one you want to take a look, I am curious too why this function doesn't work (or may not be called) during automatic restore.


There are several other major functions in PersistentWindowProcessor.cs
Start()       // kickstart timers for capture/restore
WinEventProc()  //window event handler
CaptureWindowOfInterest() // collect a list of interested windows (including taskbar)

**Reply by daktfi** [reply]
Sorry for delay - took me some time to do everything right.
First of all, I must admit that I have not that much experience with any of Visual Studio, C# or Windows (I program C++/Linux and a little bit of hobby projects in something cross-platform like Qt, still C++).
Looking at code I noticed there's a comment around RecoverTaskBarArea() saying that restoring height of horizontal taskbar is TODO. Probably that's the main reason that taskbar's height isn't restored.
Another thing (but I'm not sure about this one, as I said, not much experience with both C# and Windows): in that same RecoverTaskBarArea() there's a loop over all displays at line 2391 of PersistentWindowProcessor.cs that sets single variable "left_edge" if taskbar is on the left, opposite being taskbar on the right. However, this doesn't cover cases when taskbar is horizontal on top/bottom, and, more important, it seems it doesn't cover case when taskbars are placed differently on different screens.
I'll say that again: I don't have much experience with C#/Window and maybe there's some very specific calling or whatever and that's perfectly fine (like separate instance of RecoverTaskBarArea()'s parent class for each taskbar), but to me it _seems_ it's not. I think it's worth to make an enum of taskbar position (left/right/top/bottom) and store taskbar position for each display separately (in array). Don't know if they have different handlers, but looking at code it seems that it's possible to get correct taskbar for each display already.

Hope this help. Unfortunately I can't do much more at the moment (some family issues and forthcoming holidays, that kind of things). I'll see if I can get some quick-n-dirty C# manuals and tweak your code, but don't think that'll end up good...

**Reply by kangyu-california** [reply]
Thanks for looking into RecoverTaskBarArea(), although the real issue is why MoveTaskBar() fails to restore vertical taskbar back to horizontal position.
As I can not reproduce the issue on my setup, if you could manage to set a breakpoint in MoveTaskBar() and figure out something, It would greatly speedup the progress.

**Reply by daktfi** [reply]
Ok, I'll try to. Not much time for that, unfortunately, but I'll manage this week.

**Reply by daktfi** [reply]
Sorry, hadn't enough time that week, planned to look at this on the holidays, but my notebook decided to upgrade to Windows 11 and alas, there no more need for extra tools, it finally restores windows back to the screen they were taken from!
Anyway, many thanks for this great program and all the best wishes in New Year!

**Comment by amskris** [comment]
Thank you so much! I was pulling my hair out trying to figure out the issue. I tried all of the solution suggested on the forums (updating drivers, registry entries etc.).
